Check contracts for payment acceptances

aspiers opened the pull request:
If a contract has no transitions which accept payment, raise a warning.  **Unfortunately this introduces a lot of new warnings for existing test cases, therefore most likely the usefulness of this feature, or at least its UI, will need to be reconsidered.  Guidance from Zilliqa is needed on this point.**

Also if a transition has any code path with potentially multiple `accept` statements in it, raise a warning.  We do not raise an error since it is not possible via static analysis to know for sure in all cases whether multiple `accept` statements would be reached if present, since this can be dependent on conditions which are only known at run-time.  As a silly example, this transition may or may not attempt to accept twice:

    transition donate_once_or_twice(twice : Bool)
      accept;
      match twice with
      | True => accept
      | False =>
        e = { _eventname : "Thanks" };
        event e
      end
    end

In contrast, this transition is guaranteed to accept only once:

    transition donate_once(switch : Bool)
      match switch with
      | True => accept
      | False =>
        e = { _eventname : "Thanks" };
        event e
      end;
      match switch with
      | False => accept
      | True =>
        e = { _eventname : "Thanks" };
        event e
      end
    end

however it would be very difficult to implement analysis which catches cases like this.

vaivaswatha commented:
@aspiers: @jjcnn has a few changes recommendedin the `aspiers-check-duplicate-accepts`  on the Scilla repo. Can you update your PR by merging these?

https://github.com/aspiers/scilla/compare/check-duplicate-accepts...Zilliqa:aspiers-check-duplicate-accepts

aspiers commented:
[**@vaivaswatha**](https://github.com/vaivaswatha) commented on [February 4, 2019 1:56 PM](https://github.com/Zilliqa/scilla/pull/409#issuecomment-460258184 "2019-02-04T13:56:41Z - Replied by Github Reply Comments"):
> [@aspiers](https://github.com/aspiers): [@jjcnn](https://github.com/jjcnn) has a few changes recommendedin the `aspiers-check-duplicate-accepts` on the Scilla repo. Can you update your PR by merging these?

Done.  I'm still not convinced that the `had no transitions with accept statement` warning is worth having (it seems too noisy to me), but I'll defer to your better judgement on that ;-)

aspiers commented:
Rebased again against latest `master`, fix conflicts and tests.

jjcnn commented on src/lang/checkers/SanityChecker.ml:
```diff
@@ -69,12 +77,104 @@ module ScillaSanityChecker
+            match List.length group with
+            | 0 -> ()
+            | 1 -> ()
+            | _ -> dup_accept_warning group)
```
`List.length` traverses the entire list, so it takes time proportional to the length of the list. Since this check is equivalent to testing whether the length is greater than 1, you can instead do the following:

```
match group with
| _ :: _ :: _ -> dup_accept_warning group
| _ -> ()
```

Or you can introduce a helper function `list_empty : 'a List -> bool`.

aspiers replied:
Fixed via `match`.  BTW I notice that `batteries` provides `BatList.is_empty` so presumably that would have been another option?

jjcnn replied:
Batlist.is_empty does the same thing. I am just not used to using Batteries.

jjcnn commented on src/lang/checkers/SanityChecker.ml:
```diff
@@ -69,12 +77,104 @@ module ScillaSanityChecker
+           [] contr.ctrans)
+      in
+
+      (if List.length all_accept_groups == 0 then
```
Same story as before - you don't want to traverse the entire list just to check whether the list is empty.

Use 

```
match all_accept_groups with
| [] -> ...
| _ -> ...
```
instead.

aspiers replied:
Fixed.

jjcnn commented on src/lang/checkers/SanityChecker.ml:
```diff
@@ -69,12 +77,104 @@ module ScillaSanityChecker
     (* No repeating transition names. *)
     let e = e @ check_duplicate_ident SR.get_loc (List.map (fun t -> t.tname) contr.ctrans) in
+    (* No repeating accept statements *)
+    let e = e @ check_accepts contr in
```
This call should be guarded by the `-cf` command-line option, which enables cashflow analysis. See `src/runners/scilla_checker.ml` for an example of how this is done (you may need to pass an appropriate flag to `SanityCheckers.ml`.

aspiers replied:
Need to make sure I understand correctly: do you mean the check should be skipped if `-cf` is specified, or if it isn't specified, or something else?  I'm not yet aware of the relationship between my check and cashflow analysis.

Thanks for the implementation tip - I assume you are referring to this code in `scilla_checker.ml`:

    let cf_info_opt = if cli.cf_flag then Some (check_cashflow typed_cmod) else None in

jjcnn replied:
That is indeed the code snippet.

`-cf` activates cashflow analysis, and I think the acceptance checks fits nicely into a generic "suite of cash-related checks". So the check should be enabled when the `-cf` option is specified, but not when it is not.

aspiers replied:
Ah OK of course, that makes perfect sense - thanks!

aspiers replied:
Done.

jjcnn commented on src/lang/checkers/SanityChecker.ml:
```diff
@@ -69,12 +77,104 @@ module ScillaSanityChecker
+                (get_id transition.tname) ^
+                String.concat ""
+                  (List.map
+                     (fun loc -> Core.sprintf "  Accept at %s\n" (get_loc_str loc))
```
I wonder if the error messages could not be improved.

We don't expect there to be many `accept` statements in a code path, but maybe the best thing to do is use `warn2` which takes both a start location and an end location. You can then pass the location of either the first and second `accept`, or the location of the first and the last `accept` in the code path.

This may also simplify the analysis, since you won't need to keep track of more than 2 locations in each code path.

aspiers replied:
I think I previously tried to use `warn2` but then couldn't see any way of getting the end location.  Having said that, in this case `accept` is always 6 characters so I guess adding 6 to the start column would be good enough.

But IIUC here you are suggesting something different, i.e. to make the start location point to the beginning of one `accept`, and the end location point to the beginning of another - right?  I can think of two potential objections to that:

1. It would seem slightly inconsistent with all other warnings/errors where the expression causing the warning/error is contained within the start and end locations.  IIUC with your suggestion there would be a culprit expression *starting* at the warning's end location.

2. I deliberately decided to collect *all* instances of `accept` within each code path rather than just the minimum of 2 required to show a warning, so that if there were more than 2, the Scilla developer would be able to check (and potentially fix) them all in one go.  OTOH if the warning only shows 2 but there are more than 2, the developer would have to eliminate one of those in order to check (and potentially fix) the others.  And maybe that elimination would actually need to be reverted after checking the others, since the `accept` eliminated might not be a true duplicate for reasons which a human can deduce but which static analysis cannot (yet) infer (as per the example in this PR's description).  Worse, the developer wouldn't even be aware that there were more than 2 potential `accepts` in the same code path, so it might be that a non-duplicate warning could actually prevent any visible warning of a genuine duplicate.

The first of these two is more of an aesthetic inconsistency which could quite easily be ignored; however the second seems to infringe on the correctness of the check, which concerns me much more.

Having said that, I'm hopelessly new to all of this, so of course I defer to your expertise ;-)

jjcnn replied:
Ad. 1: That is indeed my suggestion. I don't see a huge problem in having a the end position of a warning be at the start of the violating `accept` statement, even if it is slightly inconsistent.

Ad. 2: When the typechecker discovers an error in a transition, we bail out of that transition regardless of whether there may be more (related or unrelated) type errors in the same transition, so it's not entirely inconsistent to only report two of the violating `accept`s.

The issue is not only with the console output, but also with the integration with Savant IDE (https://ide-next.zilliqa.com/) or other IDEs. If the error position is given just as one of the violating statements, then the IDE won't be able to highlight the other violating statement unless it attempts to parse the error message string.

On the other hand, if it is given the positions of the first and last violating statements, then a naïve IDE would highlight the entire code block from the first to the last violating statement, and this would most likely include all the other violating statements in the transition.

aspiers replied:
Thanks for the great explanations - I have a much better understanding now of where you're coming from.

Ad. 1: We could even have the end position be 6 characters after the start of the last `accept` in the warned codepath, so that the warning range includes every `accept` which it refers to.

Ad. 2: Points taken about the (lack of) inconsistency and the IDE highlighting.

I think the best approach which takes all these points into account is my suggestion immediately above in this comment (Ad. 1) coupled with a warning message which includes *all* potential duplicate `accepts` in the code path.  As [previously mentioned](https://github.com/Zilliqa/scilla/pull/409#discussion_r254678601), this last detail is necessary in order to maintain the correctness of the check, since if only 2 `accept` statements are mentioned in the warning and the developer decides that those are not worth worrying about, they might surmise that there are no other `accepts` in the same code path, which might not be the case.

aspiers replied:
OK, I've now done this.

aspiers commented:
I've revamped this quite a bit and pushed a new version:

- Split the `accept` checks out into `Accept.ml`
- Set the warning's end position to the end of the final `accept` in the code path
- Kept the warning message as it was, i.e. still mentioning of *all* occurrences of `accept` in order to ensure correctness of the check as discussed
- Discovered a bug in the check's traversal of the syntax tree, and fixed it separately in the second commit of this PR, to make it easier for you to review this important change in the algorithm independently of everything else
- Rebased against latest `master`

aspiers commented:
In case you read the above comment from an email notification, please re-read from the GitHub UI, since I just made some important edits to it :)
